tests/basic_test.go

```diff
@@ -439,6 +443,79 @@ func testGenerateUniqueTimestamps(t *testing.T) {
 	})
 }
 
+func testFailureLexer(t *testing.T) {
+	t.Run("with lexer error", func(t *testing.T) {
+		stateData := []byte(`routes:
+  running:
+  - destination: 0.0.0.0/0
+    next-hop-address: 192.168.100.1
+    next-hop-interface: eth1
+    table-id: 254
+`)
+
+		policySpec := types.PolicySpec{
+			Capture: map[string]string{
+				"cap1": `routes.running.destination==-"0.0.0.0/0"`,
+			},
+			DesiredState: stateData,
+		}
+		_, err := nmpolicy.GenerateState(policySpec, stateData, types.CachedState{})
+		assert.EqualError(t, err, `failed to generate state, err: failed to resolve capture expression, err: illegal rune -
+| routes.running.destination==-"0.0.0.0/0"
+| ............................^`)
+	})
+}
+
+func testFailureParser(t *testing.T) {
+	t.Run("with lexer error", func(t *testing.T) {
+		stateData := []byte(`routes:
+  running:
+  - destination: 0.0.0.0/0
+    next-hop-address: 192.168.100.1
+    next-hop-interface: eth1
+    table-id: 254
+`)
+
+		policySpec := types.PolicySpec{
+			Capture: map[string]string{
+				"cap1": `routes.running.destination=="0.0.0.0/0" |`,
+			},
+			DesiredState: stateData,
+		}
+		_, err := nmpolicy.GenerateState(policySpec, stateData, types.CachedState{})
+		assert.EqualError(t, err,
+			"failed to generate state, err: failed to resolve capture expression, "+
+				"err: invalid pipe: only paths can be piped in"+`
+| routes.running.destination=="0.0.0.0/0" |
+| ........................................^`)
+	})
+}
+
+func testFailureResolver(t *testing.T) {
+	t.Run("with lexer error", func(t *testing.T) {
+		stateData := []byte(`routes:
+  running:
+  - destination: 0.0.0.0/0
+    next-hop-address: 192.168.100.1
+    next-hop-interface: eth1
+    table-id: 254
+`)
+
+		policySpec := types.PolicySpec{
+			Capture: map[string]string{
+				"cap1": `interfaces | routes.running.destination=="0.0.0.0/0"`,
+			},
+			DesiredState: stateData,
+		}
+		_, err := nmpolicy.GenerateState(policySpec, stateData, types.CachedState{})
+		assert.EqualError(t, err,
+			"failed to generate state, err: failed to resolve capture expression, "+
+				"err: resolve error: eqfilter error: invalid path input source, only capture reference is supported"+`
+| interfaces | routes.running.destination=="0.0.0.0/0"
+| .......................................^`)
```

Comment:
in this specific case it would be more readable if the `^` would point to the `interfaces` or to the pipe. We can fix it in a follow-up.

Reply:
We will fix it at follow up after merging the PR that use a visitor for path.go

Reply:
Added the visitor part at the issue that covers the path.go refactoring https://github.com/nmstate/nmpolicy/issues/50.